[SPARK-23674][ML] Adds Spark ML Events to Instrumentation

HyukjinKwon opened the pull request:
## What changes were proposed in this pull request?

This PR proposes to add ML events to Instrumentation, and use it in Pipeline so that other developers can track and add some actions for them.

## Introduction

ML events (like SQL events) can be quite useful when people want to track and make some actions for corresponding ML operations. For instance, I have been working on integrating 
Apache Spark with [Apache Atlas](https://atlas.apache.org/QuickStart.html). With some custom changes with this PR, I can visualise ML pipeline as below:

![spark_ml_streaming_lineage](https://user-images.githubusercontent.com/6477701/49682779-394bca80-faf5-11e8-85b8-5fae28b784b3.png)

I would like to note that we can also interact this with other SQL/Streaming events in Apache Spark. For instance, where the input `Dataset` is originated. For instance, with current Apache Spark, I can visualise SQL operations as below:

![screen shot 2018-12-10 at 9 41 36 am](https://user-images.githubusercontent.com/6477701/49706269-d9bdfe00-fc5f-11e8-943a-3309d1856ba5.png)

I think we can combine those existing lineages together to easily understand where the data comes and goes. Currently, ML side is a hole so the lineages can't be connected for the current Apache Spark ..

I think it's not to mention how useful it is to track the SQL/Streaming operations. I would like to propose ML events as well (as lowest stability `@Unstable` APIs for now - no guarantee about stability).

## Implementation Details

### Event Class Implementation

**`mllib/src/main/scala/org/apache/spark/ml/events.scala`**

```scala
@Unstable
case class ...StartEvent(caller, input)
@Unstable
case class ...EndEvent(caller, output)

trait MLEvents {
  // Wrappers to send events:
  // def with...Event(body) = {
  //   body()
  //   SparkContext.getOrCreate().listenerBus.post(event)
  // }
}
```

This trait is used by `Instrumentation`.

```scala
class Instrumentation ... with MLEvents {
```

and used as below:

```scala
instrumented { instr =>
  instr.with...Event(...) {
    ...
  }
}
```

### Sending Side Implemnetation

This way mimics both:

**1. Catalog events (see `org/apache/spark/sql/catalyst/catalog/events.scala`)**

- This allows a Catalog specific listener to be added `ExternalCatalogEventListener` 

- It's implemented in a way of wrapping whole `ExternalCatalog` named `ExternalCatalogWithListener`
which delegates the operations to `ExternalCatalog`

This is not quite possible in this case because most of instances (like `Pipeline`) will be directly created by users in most of cases. We might be able to do that via extending `ListenerBus` for all possible instances but IMHO it's too invasive. Also, exposing another ML specific listener sounds a bit too much at this stage. Therefore, I simply borrowed file name and structures here

**2. SQL execution events (see `org/apache/spark/sql/execution/SQLExecution.scala`)**

- Add an object that wraps a body to send events

Current apporach is rather close to this. It has a `with...` wrapper to send events. I borrowed this approach to be consistent.

## Usage

It needs a custom implementation for a query listener. For instance,

with the custom listener below:

```scala
class CustomMLListener extends SparkListener
  def onOtherEvents(e) = e match {
    case e: MLEvent => // do something
    case _ => // pass
  }
}
```

There are two (existing) ways to use this.

```scala
spark.sparkContext.addSparkListener(new CustomMLListener)
```

```bash
spark-submit ...\
  --conf spark.extraListeners=CustomMLListener\
  ...
```

It's also similar with other existing implementation in SQL side.

## Target users

1. At least, I can see some interests going on outside.

    - SQL Listener
      - https://stackoverflow.com/questions/46409339/spark-listener-to-an-sql-query
      - http://apache-spark-user-list.1001560.n3.nabble.com/spark-sql-Custom-Query-Execution-listener-via-conf-properties-td30979.html

    - Streaming Query Listener
      - https://jhui.github.io/2017/01/15/Apache-Spark-Streaming/
      -  http://apache-spark-developers-list.1001551.n3.nabble.com/Structured-Streaming-with-Watermark-td25413.html#a25416

2. Apache Atlas and Spark users. The plugin mirror is exposed at [spark-atlas-connector](https://github.com/hortonworks-spark/spark-atlas-connector) so that anyone could do something about lineage and governance in Atlas.

## How was this patch tested?

Manually tested and unit tests were added.

SparkQA commented:
**[Test build #99868 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/99868/testReport)** for PR 23263 at commit [`a9112f3`](https://github.com/apache/spark/commit/a9112f33ff8fbfb66bad76bff6898abdef5b6881).

HyukjinKwon commented on mllib/src/main/scala/org/apache/spark/ml/Predictor.scala:
```diff
@@ -210,7 +214,7 @@ abstract class PredictionModel[FeaturesType, M <: PredictionModel[FeaturesType,
   }
 
-  protected def transformImpl(dataset: Dataset[_]): DataFrame = {
+  override protected def transformImpl(dataset: Dataset[_]): DataFrame = {
```
`transformImpl` for some abstraction and `saveImpl` are already existent.

HyukjinKwon commented:
cc @srowen, @cloud-fan (since it mimics SQL's event listener), @jkbradley, @mengxr and @yanboliang. Mind if I ask to take a look please? WDYT about this?

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/5885/
Test PASSed.

HyukjinKwon commented:
The tests pass in my local. I'll fix them shortly.

viirya commented on mllib/src/main/scala/org/apache/spark/ml/Pipeline.scala:
```diff
@@ -132,7 +132,8 @@ class Pipeline @Since("1.4.0") (
    */
   @Since("2.0.0")
-  override def fit(dataset: Dataset[_]): PipelineModel = {
+  override def fit(dataset: Dataset[_]): PipelineModel = super.fit(dataset)
```
Is there any `fit` method which doesn't do `super.fit()`?

HyukjinKwon replied:
Ah, it's there just only to keep the `@Since`. Looks some classes don't explicitly note that so I didn't call `super` in other places.

viirya commented on mllib/src/test/scala/org/apache/spark/ml/MLEventsSuite.scala:
```diff
@@ -0,0 +1,199 @@
+      .getOrCreate()
+
+    checkpointDir = Utils.createDirectory(tempDir.getCanonicalPath, "checkpoints").toString
+    sc.setCheckpointDir(checkpointDir)
```
I may miss it, where do we use checkpoint?

HyukjinKwon replied:
Let me double check and address this while fixing the test. I just copied this from `MLlibTestSparkContext`.

viirya commented on mllib/src/main/scala/org/apache/spark/ml/Estimator.scala:
```diff
@@ -65,7 +65,19 @@ abstract class Estimator[M <: Model[M]] extends PipelineStage {
+  @Since("3.0.0")
+  protected def fitImpl(dataset: Dataset[_]): M = {
+    // Keep this default body for backward compatibility.
+    throw new UnsupportedOperationException("fitImpl is not implemented.")
```
For current change, Spark ML developers can still choose to override `fit` instead `fitImpl` so their ML model can work without ML event?

HyukjinKwon replied:
Yes, that was my intention. I wanted to force to implement `fitImpl` but was thinking that might be too breaking change (it's going to at least break source compatibility). I am willing to follow other suggestions - I am pretty sure you or other guys are more familiar with ML side.

SparkQA commented:
**[Test build #99872 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/99872/testReport)** for PR 23263 at commit [`4f2fda2`](https://github.com/apache/spark/commit/4f2fda2272ee7e7d62df139c2f994ef7a122bf7c).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/5889/
Test PASSed.

SparkQA commented:
**[Test build #99873 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/99873/testReport)** for PR 23263 at commit [`ba9db6e`](https://github.com/apache/spark/commit/ba9db6eb2c98a1c84f982a093ff982a030b9eab7).

SparkQA commented:
**[Test build #99872 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/99872/testReport)** for PR 23263 at commit [`4f2fda2`](https://github.com/apache/spark/commit/4f2fda2272ee7e7d62df139c2f994ef7a122bf7c).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/99872/
Test PASSed.

SparkQA commented:
**[Test build #99873 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/99873/testReport)** for PR 23263 at commit [`ba9db6e`](https://github.com/apache/spark/commit/ba9db6eb2c98a1c84f982a093ff982a030b9eab7).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/99873/
Test PASSed.

srowen commented:
My first impression is that it's a big change, which is reason for caution here.

Visualizing a workflow is nice, but Spark's Pipelines are typically pretty straightforward and linear. I could imagine producing a nicer visualization than what you get from reading the Spark UI, although of course we already have some degree of history and data there.

These are just the hooks, right? someone would have to implement something to use these events. I see the value in the API to some degree, but with no concrete implementation, does it add anything for Spark users out of the box?

It seems like the history this generates would belong in the history server, although that already has a pretty particular purpose, storing granular history of events in Spark. Is that what someone would likely do? or would someone likely have to run Atlas to use this? If that's a good example of the use case, and Atlas is really about lineage and governance, is that the thrust of this change, to help with something to do with model lineage and reproducibility?

It's good that the API changes little, though it does change a bit.

I think I mostly have questions right now.

HyukjinKwon commented:
> Visualizing a workflow is nice, but Spark's Pipelines are typically pretty straightforward and linear. I could imagine producing a nicer visualization than what you get from reading the Spark UI, although of course we already have some degree of history and data there.

Another good thing that might have to be considered is, that we can interact this with other SQL events. For instance, where the input `Dataset` is originated. For instance, with current Apache Spark, I can visualises SQL operations as below:

![screen shot 2018-12-10 at 9 41 36 am](https://user-images.githubusercontent.com/6477701/49706269-d9bdfe00-fc5f-11e8-943a-3309d1856ba5.png)

I think we can combine those existing lineages together to easily understand where the data comes and goes.

(BTW, I hope it doesn't sound like I'm pushing this case for my one specific case - I think this hook-like feature can be useful in many way. I currently have one explicit example to show so I'm referring my case.)

> These are just the hooks, right? someone would have to implement something to use these events. I see the value in the API to some degree, but with no concrete implementation, does it add anything for Spark users out of the box?

Yes, right. It does not add anything to Spark users out of the box. It needs a custom implementation for a query listener. For instance,

with the custom listener below:

```scala
class CustomSparkListener extends SparkListener
  def onOtherEvents(e: SparkListenerEvent) = e match {
    case e: MLEvent => // do something
    case _ => // pass
  }
```

There are two (existing) ways to use this.

```scala
spark.sparkContext.addSparkListener(new CustomSparkListener)
```

```bash
spark-submit
  --conf spark.extraListeners=CustomSparkListener\
...
```

It's also similar with other existing implementation in SQL side (catalog events described above in PR description).

One actual example that I had with SQL query listener was that I had to close one connection every time after SQL execution.

> Is that what someone would likely do? or would someone likely have to run Atlas to use this? If that's a good example of the use case, and Atlas is really about lineage and governance, is that the thrust of this change, to help with something to do with model lineage and reproducibility?

There are two reasons.

1. I think someone in general would likely utilise this feature like other event listeners. At least, I can see some interests going on outside.

- SQL Listener
  - https://stackoverflow.com/questions/46409339/spark-listener-to-an-sql-query
  - http://apache-spark-user-list.1001560.n3.nabble.com/spark-sql-Custom-Query-Execution-listener-via-conf-properties-td30979.html

- Streaming Query Listener
  - https://jhui.github.io/2017/01/15/Apache-Spark-Streaming/
  -  http://apache-spark-developers-list.1001551.n3.nabble.com/Structured-Streaming-with-Watermark-td25413.html#a25416

2. Someone would likely run this via Atlas so that someone could do something about lineage and governance, yes, as you said. Yes, I'm trying to show integrated lineages in Apache Spark but this is a missing hole. There had to be this change for this.

HyukjinKwon commented:
I wanted to avoid a big change (also considering ML side is rather less active) .. but this looked to me necessarily required changes to address this issue fully .. Still there's no breaking change at all but only one addition to the developer API.

I can maybe try to do this for few specific implementations to make the change/affected scope small experimentally. However, my impression is that it's a required change to fully address this issue. I'm still willing to follow any guidance. I would appreciate it.

felixcheung commented:
my 2c - it does seem useful, though sounds like mostly for Atlas for now.
if you can think of a less intrusive way to do this it might be easier to review...

felixcheung commented:
(oops sorry - too easy to close)

HyukjinKwon commented:
Or should I maybe only do Pipeline related instances only?

srowen commented:
I can't think of a better approach; for what it's trying to do that's about the cleanest way. I think the changes are OK.

Maybe a smaller change that is still viable is indeed to put this in Pipeline rather than the implementations. That is, you only get this if you use Pipeline, sure, but that seems like the use case. Pipeline's code can be the bit that fires off events (right?) rather than reinstrumenting all the other classes. Is that possible or missing a bigger opportunity?

HyukjinKwon commented:
Yes .. right. I could add the wrapper `MLEvents.withTransformEvent { ... }` to only Pipeline codes and only enable events support for it, for now, tentatively, and yes, that's what the Atlas support needs for now. Let me try update it to make the scope small for now in this way.

HyukjinKwon committed:
Add MLEvents to Pipeline

SparkQA commented:
**[Test build #100232 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/100232/testReport)** for PR 23263 at commit [`17de2a8`](https://github.com/apache/spark/commit/17de2a87c55183f35c8ad8c76f0d59d980ebfc47).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/6206/
Test PASSed.

HyukjinKwon commented:
I made the change smaller by only sending events at `Pipeline`s. So, it should really be conservative enough in terms of being intrusive ...

I preserved previous proposal at another branch (https://github.com/apache/spark/compare/master...HyukjinKwon:SPARK-23674-2?expand=1), which adds events support for all implementations within ML. I can propose this in the future when we see some values on it.

HyukjinKwon committed:
More complete tests

SparkQA commented:
**[Test build #100237 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/100237/testReport)** for PR 23263 at commit [`3b0caa1`](https://github.com/apache/spark/commit/3b0caa18c1bdcc7b2c9bc2e71e506f6a94827dca).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/6210/
Test PASSed.

HyukjinKwon commented:
To be absolutely conservative, I can even add a configuration switch if anyone feels so.

SparkQA commented:
**[Test build #100232 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/100232/testReport)** for PR 23263 at commit [`17de2a8`](https://github.com/apache/spark/commit/17de2a87c55183f35c8ad8c76f0d59d980ebfc47).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class TransformStart(transformer: Transformer, input: Dataset[_]) extends MLEvent`
  * `case class TransformEnd(transformer: Transformer, output: Dataset[_]) extends MLEvent`
  * `case class FitStart[M <: Model[M]](estimator: Estimator[M], dataset: Dataset[_]) extends MLEvent`
  * `case class FitEnd[M <: Model[M]](estimator: Estimator[M], model: M) extends MLEvent`
  * `case class LoadInstanceStart[T](reader: MLReader[T], path: String) extends MLEvent`
  * `case class LoadInstanceEnd[T](reader: MLReader[T], instance: T) extends MLEvent`
  * `case class SaveInstanceStart(writer: MLWriter, path: String) extends MLEvent`
  * `case class SaveInstanceEnd(writer: MLWriter, path: String) extends MLEvent`

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/100232/
Test PASSed.

SparkQA commented:
**[Test build #100237 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/100237/testReport)** for PR 23263 at commit [`3b0caa1`](https://github.com/apache/spark/commit/3b0caa18c1bdcc7b2c9bc2e71e506f6a94827dca).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/100237/
Test PASSed.

srowen reviewed (commented):
Oh I had imagined you would do things like fire events for each estimator's fit() method inside here, not just for the whole Pipeline fit. Does that get a lot of the value of your original patch? likewise for transform(), etc.

HyukjinKwon commented:
Yes .. that's what I wanted to do and the previous change is really what it needs as is ..

I can go back to the previous change. I can simply checkout, pickup commit and rebase stuff. I can't think of other better way than that..

I can also leave all the changes to estimator's fit(), transform(), etc. (like the original patch ) and don't implement them except Pipeline as well to reduce change.

I am not sure which one is easiest way for you guys to take a look.

HyukjinKwon commented:
To make Atlas work for my case, any case is fine. I just need to catch the events related with Pipeline and that's all.

srowen commented:
I guess I mean, why does the event have to be fired from within each class's implementation, in fit or fitImpl? what about firing it when Pipeline calls any fit() method, etc? You could also capture events from all the elements in the pipeline without instrumenting every implementation. It's more than this change, which looks fine, but seems like it would capture the functionality you had in the original change?

HyukjinKwon commented:
Oh! let me take a look.

HyukjinKwon committed:
Address comments

SparkQA commented:
**[Test build #100279 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/100279/testReport)** for PR 23263 at commit [`98a0a4e`](https://github.com/apache/spark/commit/98a0a4ec2042aa4dbb1d0811c217a9759f0dd218).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/6249/
Test PASSed.

HyukjinKwon committed:
nit

SparkQA commented:
**[Test build #100280 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/100280/testReport)** for PR 23263 at commit [`2143b0f`](https://github.com/apache/spark/commit/2143b0f57390620fd011ea313e2bd6b4ccdf1940).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/6250/
Test PASSed.

SparkQA commented:
**[Test build #100279 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/100279/testReport)** for PR 23263 at commit [`98a0a4e`](https://github.com/apache/spark/commit/98a0a4ec2042aa4dbb1d0811c217a9759f0dd218).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/100279/
Test PASSed.

SparkQA commented:
**[Test build #100280 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/100280/testReport)** for PR 23263 at commit [`2143b0f`](https://github.com/apache/spark/commit/2143b0f57390620fd011ea313e2bd6b4ccdf1940).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/100280/
Test PASSed.

srowen reviewed (commented):
Yeah this kind of approach looks OK. Much less change. I think it's OK to log only from Pipeline as that is probably the use case here.

HyukjinKwon commented:
Sure!

HyukjinKwon committed:
Add logs at Pipeline calls

HyukjinKwon commented on mllib/src/main/scala/org/apache/spark/ml/Pipeline.scala:
```diff
@@ -301,9 +308,11 @@ class PipelineModel private[ml] (
   @Since("2.0.0")
-  override def transform(dataset: Dataset[_]): DataFrame = {
+  override def transform(dataset: Dataset[_]): DataFrame = MLEvents.withTransformEvent(
+      this, dataset, logging = true) {
```
I manually tested the logging and this logging way is borrowed from `transformSchema`.

SparkQA commented:
**[Test build #100297 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/100297/testReport)** for PR 23263 at commit [`81e2ca2`](https://github.com/apache/spark/commit/81e2ca2e17d23e8350f8ce76aaa3fef4fb755878).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/6266/
Test PASSed.

SparkQA commented:
**[Test build #100297 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/100297/testReport)** for PR 23263 at commit [`81e2ca2`](https://github.com/apache/spark/commit/81e2ca2e17d23e8350f8ce76aaa3fef4fb755878).
 * This patch **fails to generate documentation**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class DefaultParamsWriter(instance: Params) extends MLWriter `

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/100297/
Test FAILed.

HyukjinKwon commented:
retest this please

HyukjinKwon committed:
nit

SparkQA commented:
**[Test build #100306 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/100306/testReport)** for PR 23263 at commit [`2af4231`](https://github.com/apache/spark/commit/2af42313bcaa1f127aa8b0c9751b37a613856fe2).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/6275/
Test PASSed.

SparkQA commented:
**[Test build #100306 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/100306/testReport)** for PR 23263 at commit [`2af4231`](https://github.com/apache/spark/commit/2af42313bcaa1f127aa8b0c9751b37a613856fe2).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/100306/
Test PASSed.

mengxr commented:
@HyukjinKwon Have we considered integrating with ML instrumentation?
 https://github.com/apache/spark/blob/master/mllib/src/main/scala/org/apache/spark/ml/util/Instrumentation.scala#L181

They serve similar purpose.

HyukjinKwon commented:
I'm not used to ML side so it needs a while to check .. Let me check and be back soon!

HyukjinKwon commented:
@mengxr, looks actually considered at here https://github.com/apache/spark/pull/23263#issuecomment-447885870.

I also checked by myself and it looks at least requiring a considerable change to update each implementation similarly with the previous way.

HyukjinKwon commented:
Like, if we fix 

https://github.com/apache/spark/blob/master/mllib/src/main/scala/org/apache/spark/ml/util/Instrumentation.scala#L181

we should touch each implementation that uses it.

If we add another method similarly with it, we should also touch each implementation to wrap it.

These will virtually require similar changes as I first proposed - https://github.com/apache/spark/compare/master...HyukjinKwon:SPARK-23674-2?expand=1

cloud-fan commented:
AFAIK both the Spark core UI and Spark SQL UI are implemented via watching the events. I feel it's OK to do the same for MLlib, but the tricky part is we don't have built-in Mllib UI feature...

srowen commented:
That's a good point then, what to do with the ML instrumentation? is it deprecated as a old version of 'events'? is it still used? it seems funny to have two event logging systems here, but could the old instrumentation be removed? I am not sure who downstream might be using it.

HyukjinKwon commented:
Let me give a shot to make it with ML instrumentation to show how it's going to look like.

viirya commented:
ML instrumentation defines more details in a training session. Although there are few things overlapping with this ML events, some others are not. So I think instrumentation might not be removed because adding ML events.

HyukjinKwon committed:
Integrate MLEvents to Instrumentation

HyukjinKwon committed:
nit

SparkQA commented:
**[Test build #100414 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/100414/testReport)** for PR 23263 at commit [`7b26c6e`](https://github.com/apache/spark/commit/7b26c6ee27c06c3529d8d9724d440c4e6399b41d).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/6378/
Test PASSed.

SparkQA commented:
**[Test build #100414 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/100414/testReport)** for PR 23263 at commit [`7b26c6e`](https://github.com/apache/spark/commit/7b26c6ee27c06c3529d8d9724d440c4e6399b41d).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/100414/
Test FAILed.

HyukjinKwon committed:
Add comments and fix nits

SparkQA commented:
**[Test build #100416 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/100416/testReport)** for PR 23263 at commit [`bf65366`](https://github.com/apache/spark/commit/bf65366c3d3a8736ab7162df614bff11bb6bcc5d).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/6380/
Test PASSed.

HyukjinKwon committed:
Don't forget to log

HyukjinKwon commented:
@mengxr, @srowen and @felixcheung, I think I managed to integrate and address all the concerns and comments.

SparkQA commented:
**[Test build #100417 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/100417/testReport)** for PR 23263 at commit [`04954e7`](https://github.com/apache/spark/commit/04954e7bb0d616f0dcb5b967b083016fee314c03).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/6381/
Test PASSed.

SparkQA commented:
**[Test build #100416 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/100416/testReport)** for PR 23263 at commit [`bf65366`](https://github.com/apache/spark/commit/bf65366c3d3a8736ab7162df614bff11bb6bcc5d).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/100416/
Test PASSed.

SparkQA commented:
**[Test build #100417 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/100417/testReport)** for PR 23263 at commit [`04954e7`](https://github.com/apache/spark/commit/04954e7bb0d616f0dcb5b967b083016fee314c03).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/100417/
Test PASSed.

HyukjinKwon commented:
retest this please

SparkQA commented:
**[Test build #100476 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/100476/testReport)** for PR 23263 at commit [`04954e7`](https://github.com/apache/spark/commit/04954e7bb0d616f0dcb5b967b083016fee314c03).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/6429/
Test PASSed.

SparkQA commented:
**[Test build #100476 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/100476/testReport)** for PR 23263 at commit [`04954e7`](https://github.com/apache/spark/commit/04954e7bb0d616f0dcb5b967b083016fee314c03).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/100476/
Test PASSed.

HyukjinKwon commented:
Hey, @mengxr, can you check if your comment is addressed?

HyukjinKwon commented:
gentle ping @mengxr. I think the comment is addressed .. I know it's been a holiday period so everybody isn't quite available :). but mind if I ask to take a look please?

HyukjinKwon commented:
Gentle ping @mengxr

HyukjinKwon commented:
retest this please

SparkQA commented:
**[Test build #101050 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/101050/testReport)** for PR 23263 at commit [`04954e7`](https://github.com/apache/spark/commit/04954e7bb0d616f0dcb5b967b083016fee314c03).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/6918/
Test PASSed.

SparkQA commented:
**[Test build #101050 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/101050/testReport)** for PR 23263 at commit [`04954e7`](https://github.com/apache/spark/commit/04954e7bb0d616f0dcb5b967b083016fee314c03).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/101050/
Test FAILed.

SparkQA commented:
**[Test build #101061 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/101061/testReport)** for PR 23263 at commit [`2f6c9f7`](https://github.com/apache/spark/commit/2f6c9f76f1d8153a006fb2c0522ac3a3ba248c5d).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/6929/
Test PASSed.

HyukjinKwon committed:
Upgrade Mockito to 2.23.4

HyukjinKwon commented:
@mengxr, can you take a look please when you're available?

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/6931/
Test PASSed.

SparkQA commented:
**[Test build #101063 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/101063/testReport)** for PR 23263 at commit [`d27242b`](https://github.com/apache/spark/commit/d27242bf88e2c68a42659c71620302e5e4afdf6d).

SparkQA commented:
**[Test build #101063 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/101063/testReport)** for PR 23263 at commit [`d27242b`](https://github.com/apache/spark/commit/d27242bf88e2c68a42659c71620302e5e4afdf6d).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/101063/
Test PASSed.

HyukjinKwon commented:
@mengxr ping ..

HyukjinKwon commented:
Hey all, do you mind if I ask to take another look? I think I addressed all comments, and the changes here are not invasive at all. Doesn't change any previous behaviour but purely incremental improvement. Changes are quite isolated. Looks there are positive feedback in general about this feature.

srowen commented:
It's not a big change, yeah. One concern had been whether this duplicated the existing ML instrumentation. I see that this uses the same instrumentation mechanism now, right? That's good. The current instrumentation generally just logs info to the log. The new MLEvents don't do anything by default, if I read this correctly. Would it make sense to cause them to log messages by default?

srowen commented on mllib/src/main/scala/org/apache/spark/ml/Pipeline.scala:
```diff
@@ -197,10 +200,12 @@ object Pipeline extends MLReadable[Pipeline] {
     SharedReadWrite.validateStages(instance.getStages)
 
+    override def save(path: String): Unit =
+      instrumented(_.withSaveInstanceEvent(this, path, logging = true)(super.save(path)))
```
I see, yes I think it's good to emit log statements by default to match `Instrumentation`, and think it's always fine to emit a debug log. I had imagined that maybe a default implementation of `MLEvents` calls `logInfo` or similar for these events, and then anyone can provide a different implementation. Or, always logging at debug level no matter what the implementation also seems OK.

HyukjinKwon replied:
Ah, I see. Maybe emitting debugging logs alone should be fine. Developers can do, for instance, 

```scala
instrumented { inst => instr.withFitEvent(estimator, dataset) {
    instr.logDataset(dataset)
    ...
  }
}
```

```scala
instrumented { instr =>
  val output = instr.withTransformEvent(transformer, cur) {
    ...
  }
  instr.logError(...)
  instr.logDataset(output)
}
```

HyukjinKwon replied:
but let me try to deal with it. I think I can do this with minimised changes.

HyukjinKwon committed:
Emits a debug-level log by default, and poll `logEvent` out

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/7188/
Test PASSed.

SparkQA commented:
**[Test build #101371 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/101371/testReport)** for PR 23263 at commit [`8382d2c`](https://github.com/apache/spark/commit/8382d2c6272afc918efb7b93e08836942669834c).

HyukjinKwon committed:
Make the code shorter

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/7190/
Test PASSed.

SparkQA commented:
**[Test build #101371 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/101371/testReport)** for PR 23263 at commit [`8382d2c`](https://github.com/apache/spark/commit/8382d2c6272afc918efb7b93e08836942669834c).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/101371/
Test PASSed.

SparkQA commented:
**[Test build #101373 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/101373/testReport)** for PR 23263 at commit [`d79d9b2`](https://github.com/apache/spark/commit/d79d9b2c938c3aacdb6946dab3f28c0a0bca7404).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/101373/
Test PASSed.

HyukjinKwon commented:
Hey all, gentle ping ...

srowen reviewed (approved):
It's OK by me.

merlintang commented:
great work @HyukjinKwon

viirya commented on mllib/src/test/scala/org/apache/spark/ml/MLEventsSuite.scala:
```diff
@@ -0,0 +1,255 @@
+            assert(e.path.endsWith("writableStage"))
+          case e: SaveInstanceEnd if e.writer.isInstanceOf[DefaultParamsWriter] =>
+            assert(e.path.endsWith("writableStage"))
+          case e: SaveInstanceStart if getInstance(e.writer).isInstanceOf[PipelineModel] =>
```
hmm, what difference between this `SaveInstanceStart` and above one?

viirya commented on mllib/src/main/scala/org/apache/spark/ml/Pipeline.scala:
```diff
@@ -132,7 +133,8 @@ class Pipeline @Since("1.4.0") (
    * @return fitted pipeline
```
Where should we document the ml events? Should we add doc at places we add ml events like `Pipeline.fit`? Or write some doc in ml document?

viirya replied:
Ok. I see.

felixcheung commented:
@mengxr are you ok with this?

if we don't hear back in a few more days I think we should go ahead. looking much cleaner, great work @HyukjinKwon

HyukjinKwon commented:
Thank you all so much. Could we maybe go forward?

srowen commented:
I'd say go ahead today; don't think you'll get more feedback at this point.

HyukjinKwon commented:
Merged to master.

dongjoon-hyun commented:
Finally, great, @HyukjinKwon !
